include/eve/module/complex/complex.hpp

```diff
@@ -214,9 +214,9 @@ namespace eve
     {
       auto [a, b] = self;
       auto [c, d] = o;
-      real(self) = diff_of_prod(a, c, b, d);
-      imag(self) = sum_of_prod(a, d, b, c);
-      return self;
+      real(self) = a*c-b*d;
+      imag(self) = a*d+b*c;
```

Comment:
Curious: what happened with those functions?

Reply:
in fact the purpose of all that is to change the behaviour of sum_of_prod and diff_of_prod to always call fma functions and not the naive calls. This is done.
But it implies that the compl test/unit/module/complex/mul.cpp
ex multiplication is in all case 2 times (at least) slower than the naive call.
So the naive call is used in complex multiplication and sum of prod diff of prod in pedantic.
The use of these functions implies  slower execution but also better accuracy. The part added after is any(is_not_finite...
are to provide some amelioration in corner cases as (1+inf i)*(1+0i)

Reply:
I didn't get all of that but:

Does this mean that `sum_of_prod` and `diff_of_prod` are pedantic by default?
This seems weird, since most things we have a fast by default, no?

As a user I'd expect that I should always call a function with a proper name, not that I shlould do it by hand in some cases.

Reply:
Basically, if `diff_of_prod` is worse than `a*c-b*d` - we have a problem.

Reply:
the problem with diff_of_prod is that contrarily to fma if fma is not available it is lot worse than ab-cd. 
If hardware fma is there the worse factor is only 2 and it presents accuracy advantage (no more than 3 ulps from correct roundng)
So diff_of prod must be used only if needed. 
With the solution implemented there, up to now it is  used once or twice in huge special functions. And no more in complex  
The new version of the documentation will put its description  in a section "accuracy helpers" that will also contains functions as two_prod and two_add that are meant to help emulate fma when the hardare lacks.

Reply:
accuracy and speed are two separate but related problems.
Perhaps a compilation option style -DACCURATE_MATH could be used, but it seems more complicated.

Reply:
No - the switches are not the solution I think.

I see your point but I feel ify about it, I would think that if there is a function "diff_of_prod" I should use it. And then have settings around it to do what I want.

But the notion of "accuracy helpers" also makes sense. I will leave this to you and @jfalcou

Reply:
we **may** actually have a raw(diff_of_prod) that just do a*b-c*d and have the other do the proper thing.
Then complex mul use raw(diff_of_prod) the pedantic use the regular.

THe cost of actual diff_of_prod with FMA semantic is high so we can't just use it willy nilly or our speed compared to auto vectorized complex multiplication is bad.